crates/datasources/src/excel/mod.rs

```diff
@@ -1,89 +1,109 @@
 
-fn infer_value_type(v: &calamine::Data) -> Result<DataType, Error> {
-    match v {
-        calamine::Data::Int(_) => Ok(DataType::Int64),
-        calamine::Data::Float(_) => Ok(DataType::Float64),
-        calamine::Data::Bool(_) => Ok(DataType::Boolean),
-        calamine::Data::String(_) => Ok(DataType::Utf8),
-        calamine::Data::Error(e) => Err(Error::Load { msg: e.to_string() }),
-        calamine::Data::DateTime(_) => Ok(DataType::Date64),
-        calamine::Data::Empty => Ok(DataType::Null),
-        _ => Err(Error::Load {
-            msg: "Failed to parse the cell value".to_owned(),
-        }),
-    }
-}
+pub mod errors;
+pub mod stream;
+pub mod table;
 
-fn infer_schema(
-    r: &Range<calamine::Data>,
+use errors::ExcelError;
+
+pub struct ExcelTable {
+    cell_range: Range<calamine::Data>,
     has_header: bool,
-    infer_schema_length: usize,
-) -> Result<Schema, Error> {
-    let mut col_types: HashMap<&str, HashSet<DataType>> = HashMap::new();
-    let mut rows = r.rows();
-    let col_names: Vec<String> = rows
-        .next()
-        .unwrap()
-        .iter()
-        .enumerate()
-        .map(
-            |(i, c)| match (has_header, c.get_string().map(|s| s.to_string())) {
-                (true, Some(s)) => Ok(s),
-                (true, None) => Err(Error::Load {
-                    msg: "failed to parse header".to_string(),
-                }),
-                (false, _) => Ok(format!("col{}", i)),
-            },
-        )
-        .collect::<Result<_, _>>()?;
+}
 
-    for row in rows.take(infer_schema_length) {
-        for (i, col_val) in row.iter().enumerate() {
-            let col_name = col_names.get(i).unwrap();
-            let col_type = infer_value_type(col_val).unwrap();
-            if col_type == DataType::Null {
-                continue;
+impl ExcelTable {
+    pub async fn open(
+        store_access: GenericStoreAccess,
+        source_url: DatasourceUrl,
+        sheet_name: Option<&str>,
+        has_header: bool,
+    ) -> Result<ExcelTable, ExcelError> {
+        match source_url {
+            DatasourceUrl::File(path) => {
+                let path = ioutil::resolve_path(&path)?;
+                let mut sheet = calamine::open_workbook_auto(path)?;
+
+                let first_sheet = sheet_name.map(Cow::Borrowed).unwrap_or_else(|| {
+                    let sheets = sheet.sheet_names();
+                    let first = sheets.first().expect("file has a sheet");
+                    Cow::Owned(first.clone())
+                });
+
+                Ok(ExcelTable {
+                    cell_range: sheet.worksheet_range(&first_sheet)?,
+                    has_header,
+                })
+            }
+
+            DatasourceUrl::Url(_) => {
+                let store = store_access.create_store()?;
+                let list = store_access
+                    .list_globbed(&store, source_url.path().as_ref())
+                    .await?;
+
+                if list.is_empty() {
```

Comment:
should we error if there's more than one file? do we have a good way of handling multiple files?

Reply:
yup an err makes sense for now if there's more than one file, likely this bit would get refactored when adding the machinery for `create database excel`

Reply:
my initial expectation is that "create database excel" would still be only one file: each and sheet would be it's own table...

Reply:
ah... alright then.